Orange/data/table/base.py

```diff
@@ -0,0 +1,1383 @@
+import os
+import zlib
+from collections import Sequence
+from threading import Lock
+from numbers import Number
+from io import StringIO
+
+from itertools import chain
+import numpy as np
+import scipy.sparse as sp
+import pandas as pd
+import pandas.core.internals
+
+from Orange.data import Domain, StringVariable, ContinuousVariable, \
+    DiscreteVariable, Variable, TimeVariable, filter_visible
+from Orange.util import flatten, deprecated
+
+
+class _transferer:
+    """A 'minor' hack for transferring attributes to TableSeries in _constructor_sliced."""
+    def __init__(self, cls, attrs):
+        self.cls = cls
+        self.attrs = attrs
+
+    # this is a class and not a function because sometimes, pandas
+    # wants _constructor_sliced.from_array
+    def from_array(self, *args, **kwargs):
+        return self._attr_setter(self.cls.from_array(*args, **kwargs))
+
+    def __call__(self, *args, **kwargs):
+        return self._attr_setter(self.cls(*args, **kwargs))
+
+    def _attr_setter(self, target):
+        for k, v in self.attrs.items():
+            setattr(target, k, v)
+        return target
+
+
+# noinspection PyPep8Naming
+class TableBase:
+    """An abstract base class for data storage structures in Orange."""
+
+    # a list of all kwargs supported by this class,
+    # used for proper compatibility with pandas' constructors
+    _ORANGE_KWARG_NAMES = {"weights", "attributes"}
+
+    # the default name for weights columns
+    # subclasses may override this to rename the column
+    _WEIGHTS_COLUMN = "__weights__"
+
+    # a collection of all known internal column names
+    # use case: subclasses add additional columns (not only weights) and we need to know
+    # of all existing names to be able to process things wrt these internals,
+    # e.g. adding missing columns in the constructor, transparently passing them through
+    # when selecting a subset
+    _INTERNAL_COLUMN_NAMES = [_WEIGHTS_COLUMN]
+
+    # a counter for indexing rows, important for deterministically selecting rows
+    # and keeping pandas indices sane
+    _next_instance_id = 0
+    _next_instance_lock = Lock()
+
+    # the conversion cache for converting tables to different domains
+    _conversion_cache = None
+
+    # custom properties, preserved through pandas manipulations
+    _metadata = ['name',  # the table name
+                 'domain',  # the domain assigned to the table
+                 'attributes',  # any custom attributes the table has
+                 '__file__',  # if read from a file, the filename of that file
+                 '_already_inited']  # skip multiple constructor calls, see __init__
```

Comment:
Pandas has a weird way of overriding `__getitem__`, and trying to access a custom attribute that's not in metadata results in an infinite recursion.